### PR TITLE
controller: Allow controllers to exit

### PR DESCRIPTION
**Pull request by raybejjani**
_I recommend viewing the diff with whitespace ignored https://github.com/cilium/cilium/pull/2919/files?w=1_

**Summary of changes**:

In some circumstances a controller needs to encode complex logic on when
to terminate or otherwise stop. This can be mimicked by blocking, or via
more involved mechanisms utilising other goroutines. This change adds a
return type that can be handled by the controller wrapper and signals
the desire to exit. The reason given is recorded in the controller's
status.

I'm not sure that this is the best way to do this... but other schemes required a lot more coordination around channels (I tried supporting a user provided stop channel in https://github.com/cilium/cilium/compare/master...raybejjani:controller-selfstop?expand=1 but it requires custom code in the StopFunc and more complexity on the user's part). We could say that controllers should never exit and that's that but in https://github.com/cilium/cilium/pull/2772 we have a problem that controllers need to wait on other subsystems to initialise before they can know that they should just stop.

Supports: #2772 and #2893 

**How to test (optional)**:
I added a unit test.

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -173,18 +183,27 @@ func (c *Controller) runController() {
...
-				if c.params.ErrorRetryBaseDuration != time.Duration(0) {
-					interval = time.Duration(errorRetries) * c.params.ErrorRetryBaseDuration
-				} else {
-					interval = time.Duration(errorRetries) * time.Second
+			switch err := err.(type) {
+			case ExitReason:
+				// This is actually not an error case, but it causes an exit
+				c.recordSuccess()
+				c.lastError = err // This will be shown in the controller status
+				//goto shutdown
```
Comment (aanm) [inline pkg/controller/controller.go]: No shutdown?

Comment (raybejjani) [reply pkg/controller/controller.go]: oops, that was there to test my test... which apparently doesn't work. Thanks for catching it!

Comment (raybejjani) [reply pkg/controller/controller.go]: fixed :)

**Comment by raybejjani** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by ianvernon** [comment]
I was under the impression that controllers would always be running for the duration of the lifetime of an endpoint - so the fact that controllers now have the capability to exit is odd to me.

**Comment by tgraf** [comment]
> I was under the impression that controllers would always be running for the duration of the lifetime of an endpoint - so the fact that controllers now have the capability to exit is odd to me.

Controllers can be attached to objects such as endpoint via the ControllerManager. Controllers can also be used to launch an async operation which only requires retries on failure but is otherwise not attached to the lifecycle of an object.

**Pull request merged**